### PR TITLE
Refactored table creation to use the Asset_Tables_Config table

### PR DESCRIPTION
**Pull request by yardenvan**
In addition I also changed naming conventions in the AssetObject class from the previous PR.

### CODE REVIEW

File: src/main/java/com/google/cloudassets/discovery/AssetKind.java [hunk 1]
```diff
@@ -24,5 +35,50 @@ private AssetKind(String kind) {
...
+    one table name returned from the provided query it means that the Asset_Tables_Config is not
+    configured properly and therefore unexpected behaviors may arise. In this case an error is logged.
+     */
+    private static String getTableName(String queryStr, String errorMsg) {
+        ResultSet resultSet = executeStringQuery(queryStr);
+        String tableName = null;
+        if (resultSet.next()) {
+            tableName = resultSet.getString("assetTableName");
+        } else {
+            String error_msg = "No table" + errorMsg;
```
Comment (itaitz) [inline src/main/java/com/google/cloudassets/discovery/AssetKind.java]: Few comments about this:
- 'error_msg' is not according to the style guide
- Having both 'error_msg' and 'errorMsg' is very confusing.
- Why not throw an exception? This failure is probably fatal.
- Instead of passing in a partial error-message and putting it together here, it is better practice to throw and exception here, and re-throw in the caller function, wrapping this exception in another one adding more info (e.g. adding the kind)

File: src/main/java/com/google/cloudassets/discovery/Main.java [hunk 2]
```diff
@@ -70,6 +78,20 @@ private static void updateAllProjectsAssets() {
         }
     }
 
+    /*
+This function retrieves the project ids for which this process should run and returns it as a
```
Comment (itaitz) [inline src/main/java/com/google/cloudassets/discovery/Main.java]: Can you re-format this? (Also in a few places below)

File: src/main/java/com/google/cloudassets/discovery/Main.java [hunk 3]
```diff
@@ -108,44 +115,72 @@ that already existed before this process began to run (no need to delete from ta
...
+        setNewSupportedTableNames();
+        createTablesIfNotExist();
+    }
 
-        for (AssetTable table : AssetTable.values()) {
-            if (!existingTableNames.contains(table.getTableName())) {
-                tablesToCreateQueries.add(table.getTableCreateQuery());
-            }
+    /*
+    This function updates the existingTableNames variable which is a list of string that represents
```
Comment (itaitz) [inline src/main/java/com/google/cloudassets/discovery/Main.java]: "list of strings"

File: src/main/java/com/google/cloudassets/discovery/Main.java [hunk 4]
```diff
@@ -108,44 +115,72 @@ that already existed before this process began to run (no need to delete from ta
...
+            existingTableNames.add(resultSet.getString("table_name"));
         }
+    }
 
-        executeTablesCreation(tablesToCreateQueries);
+    /*
+    This function updates the newSupportedTableNames variable which is a list of string that represents
+    all of the asset tables which do not currently exist in our spanner db but were added to the
+    Asset_Tables_Config table.
+    This function relays on existingTableNames variable being properly initialized.
```
Comment (itaitz) [inline src/main/java/com/google/cloudassets/discovery/Main.java]: "relies"?

File: src/main/java/com/google/cloudassets/discovery/Main.java [hunk 5]
```diff
@@ -108,44 +115,72 @@ that already existed before this process began to run (no need to delete from ta
...
+
+        ResultSet resultSet = executeStringQuery(GET_TABLES_LIST_QUERY);
+        while (resultSet.next()) {
+            existingTableNames.add(resultSet.getString("table_name"));
         }
+    }
 
-        executeTablesCreation(tablesToCreateQueries);
+    /*
+    This function updates the newSupportedTableNames variable which is a list of string that represents
```
Comment (itaitz) [inline src/main/java/com/google/cloudassets/discovery/Main.java]: "list of strings"

File: src/main/java/com/google/cloudassets/discovery/TableCreation.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    private static final String GET_COMMON_PRIMARY_KEYS_QUERY = "SELECT columnName "
+                                                        + "FROM Asset_Tables_Config "
+                                                        + "WHERE assetTableName = 'forAllAssets' "
+                                                        + "and isPrimaryKey = True "
+                                                        + "ORDER BY primaryKeyIndex";
+
+    private static StringBuilder createStatement;
+
+    /**
+     * This function generates an DDL create table statement for the given table name.
```
Comment (itaitz) [inline src/main/java/com/google/cloudassets/discovery/TableCreation.java]: I think "This function returns a ..." is more readable

**Review by itaitz — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/cloudassets/discovery/AssetKind.java (hunk 1 above)
Comment (yardenvan) [reply src/main/java/com/google/cloudassets/discovery/AssetKind.java]: I was indeed wondering what was the best way to handle these kinds of errors:
1. The getAssetTableName & getMainTableName functions which are using this function are being called in several places in our program so throwing an error will make it necessary to handle these exceptions in the same way in different places. In order to avoid this I thought to run a validation function on our configuration tables once in the beginning of the back-end process, but this might be redundant.. WDYT?
2. Should we stop the program execution upon encountering this error? If so, what is the best way to do this?
3. Should I create a new Exception class for these errors or is this redundant?

Comment (itaitz) [reply src/main/java/com/google/cloudassets/discovery/AssetKind.java]: These are indeed non-trivial questions... Let's go over these questions over VC

Comment (yardenvan) [reply src/main/java/com/google/cloudassets/discovery/AssetKind.java]: Fixed this issue with new exceptions as we discussed (as a first step before adding system tests). 
Would love to get your input if it looks okay now.

File: src/main/java/com/google/cloudassets/discovery/AssetKind.java [hunk 7]
```diff
@@ -24,5 +33,54 @@ private AssetKind(String kind) {
...
+        String tableName = null;
+        if (resultSet.next()) {
+            tableName = resultSet.getString("assetTableName");
+        } else {
+            throw new ConfigTableException("No table" + errorMsg);
+        }
+
+        // Each asset should have exactly one table name associated with it
+        if (resultSet.next()) {
+            throw new ConfigTableException("More then one table" + errorMsg);
```
Comment (itaitz) [inline src/main/java/com/google/cloudassets/discovery/AssetKind.java]: Passing in the partial error message in this way makes the interface unnecessarily "thick" (we usually thrive for "thin" interfaces where a function does exactly one well-defined thing.) E.g. here, in addition to the main semantics (i.e. return the name of the table) there are also rather complex semantics of how the error-message sentences are being constructed.

Can you rephrase this so that 'getTableName' simply throws an exception when there is a problem, and the calling functions wrap this in their own exceptions, while adding the necessary context information in this wrapping exceptions? (Both exceptions can be 'ConfigTableException's if convenient)

File: src/main/java/com/google/cloudassets/discovery/Main.java [hunk 8]
```diff
@@ -22,46 +24,62 @@
...
         readFromDb = dbClient.readOnlyTransaction();
-        createTablesIfNotExist();
 
         // Use spanner DatabaseClient
-        updateAllProjectsAssets();
+        try {
+            maintainTables();
+            updateAllProjectsAssets();
+        } catch (TableCreationException exception) {
+            String errorMsg = "Encountered an error while running table maintenance functions.";
```
Comment (itaitz) [inline src/main/java/com/google/cloudassets/discovery/Main.java]: You can simply put the error-message literal inside the log() call to make this more idiomatic and readable.

Also - this code "swallows" the exception. Are you sure this is what we want here?
(also below)

File: src/main/java/com/google/cloudassets/discovery/TableCreation.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+                                                        + "allowCommitTimestamp "
+                                                        + "FROM Asset_Tables_Config";
+    // The order of the primary keys in the create table statement matters
+    private static final String GET_COMMON_PRIMARY_KEYS_QUERY = "SELECT columnName "
+                                                        + "FROM Asset_Tables_Config "
+                                                        + "WHERE assetTableName = 'forAllAssets' "
+                                                        + "and isPrimaryKey = True "
+                                                        + "ORDER BY primaryKeyIndex";
+
+    private static StringBuilder createStatement;
```
Comment (itaitz) [inline src/main/java/com/google/cloudassets/discovery/TableCreation.java]: Can you make this non-static? Static variables are essentially global (in the sense that all the objects share them) and we usually want to build code that can be used in parallel for different purposes (e.g. even if we don't plan to use it more than once in our system, some tests may want to instantiate it multiple times, or a future refactoring of the code may need this ability)

**Review by itaitz — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/cloudassets/discovery/Main.java (hunk 8 above)
Comment (yardenvan) [reply src/main/java/com/google/cloudassets/discovery/Main.java]: I "swallowed" the exceptions here in order to make sure that the db and spanner are properly closed even if an error is encountered. Once one of there exceptions occurs then our tables aren't really updated so the only concern is missing this failure. Do you have a better way to do this?

Comment (yardenvan) [reply src/main/java/com/google/cloudassets/discovery/Main.java]: I changed this to throw the exceptions as we discussed. However, in the end I used 'finally' instead of try-with-resources for two reasons:
1. In order to use try-with-resources the resources must be declared inside the try block - but I prefer these resources to be global
2. In between the initialisation of the 'spanner' object and the 'readFromDb' object there are two additional objects that must be initialised but they both do not implement java.io.Closeable (as they indeed do not need to be closed) and therefore can not be places in the try statment
Hope this makes sense to you

File: src/main/java/com/google/cloudassets/discovery/exceptions/ConfigTableException.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package com.google.cloudassets.discovery.exceptions;
```
Comment (itaitz) [inline src/main/java/com/google/cloudassets/discovery/exceptions/ConfigTableException.java:1]: In order to save some work and reduce boilerplate, you can create a class called "AssetDiscoveryExceptions" and put all these exception classes as static nested classes.

**Review by itaitz — APPROVED** [review]

File: src/main/java/com/google/cloudassets/discovery/exceptions/ConfigTableException.java (hunk 10 above)
Comment (yardenvan) [reply src/main/java/com/google/cloudassets/discovery/exceptions/ConfigTableException.java:1]: Good idea, will add it to the PR in which we will refactor the use of the error messages to use fluent logger string parsing
